server/etcdserver/api/v3rpc/qmon.go

```diff
@@ -0,0 +1,272 @@
+	AdmitReq(req interface{}) bool
+
+	//Stop the monitoring
+	Stop()
+}
+
+type BandwidthMonitor struct {
+	totalMemoryBudget            uint64
+	defaultRespSize              uint64
+	degradedBandwidthBytesPerSec uint64
+	burstBytes                   uint64
+	budgetExhausted              bool
+	resetTimer                   time.Duration
+	estRespSize                  *adt.CountMinSketch
+	auditThresholdPercent        uint64
+	auditOn                      bool
+	throttle                     *rate.Limiter
+	mu                           sync.Mutex
+}
+
+func NewQueryMonitor(s *etcdserver.EtcdServer) QueryMonitor {
+	var qm BandwidthMonitor
+	qm.totalMemoryBudget = DefaultTotalMemoryBudget
+	qm.defaultRespSize = DefaultRespSize
+	qm.degradedBandwidthBytesPerSec = DefaultDegradedBandwidthBytesPerSec
+	qm.burstBytes = DefaultBurstBytes
+	qm.resetTimer = DefaultResetTimer
+	qm.estRespSize, _ = adt.NewWithEstimates(0.0001, 0.9999)
+	qm.auditOn = false
+	qm.auditThresholdPercent = DefaultAuditThresholdPercent
+	qm.throttle = rate.NewLimiter(rate.Every(time.Second/time.Duration(qm.degradedBandwidthBytesPerSec)), int(qm.burstBytes))
+	qm.budgetExhausted = false
+	return &qm
+}
+
+func (ctrl *BandwidthMonitor) Start() {
+	go ctrl.start()
+}
+
+func (ctrl *BandwidthMonitor) start() {
+	ctrl.update()
+	ctrl.periodicReset()
+}
+
+func (ctrl *BandwidthMonitor) periodicReset() {
+	ticker := time.NewTicker(ctrl.resetTimer)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			ctrl.update()
+		}
+	}
+}
+
+func (ctrl *BandwidthMonitor) update() {
+	//TODO err handling
+	rss := getCurrentRssBytes()
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	ctrl.resetRespSizeUnsafe()
+	ctrl.updateBudgetUnsafe(uint64(rss))
+	ctrl.updateAuditFlagUnsafe(uint64(rss))
+}
+
+func (ctrl *BandwidthMonitor) resetRespSizeUnsafe() {
+	ctrl.estRespSize, _ = adt.NewWithEstimates(0.0001, 0.9999)
+}
+
+func (ctrl *BandwidthMonitor) updateAuditFlagUnsafe(rss uint64) {
+	ctrl.auditOn = true
+}
+
+func (ctrl *BandwidthMonitor) updateBudgetUnsafe(rss uint64) {
+	if ctrl.totalMemoryBudget <= uint64(rss) {
+		ctrl.budgetExhausted = true
+		runtime.GC()
+	} else {
+		ctrl.budgetExhausted = false
+	}
+}
+
+func (ctrl *BandwidthMonitor) isDeclinedHelper(q Query) (bool, uint64) {
+	respSize := q.qsize
+	if q.qtype == QueryTypeRange {
+		respSize = ctrl.estRespSize.EstimateString(q.qid)
+		if respSize == 0 {
+			//We have not seen a response for this type of query.
+			//Estimate based on default
+			respSize = ctrl.defaultRespSize
+		}
+		q.qsize = respSize
+	}
+	if ctrl.budgetExhausted {
+		//decline
+		return true, respSize
+	}
+
+	//admit
+	return false, respSize
+}
+
+func (ctrl *BandwidthMonitor) newQuery(req interface{}, resp interface{}) Query {
+	var reqSize, respSize int
+	var reqContent string
+	qtype := QueryTypeUnknown
+	switch _resp := resp.(type) {
+	case *pb.RangeResponse:
+		//TODO verify status is 200
+		_req, ok := req.(*pb.RangeRequest)
```

Comment:
If think the logic could contains more 'heuristics': 

https://github.com/etcd-io/etcd/blob/6af760131e1cb2181237c3eb41db2e58686c65e3/api/etcdserverpb/rpc.proto#L432

a) range_end empty -> it's a lookup -> small request.
b) limit < threshold (e.g. 100)  -> medium request.